ddtrace/tracer/sqlcomment.go

```diff
@@ -0,0 +1,135 @@
+	// SQLInjectionModeFull represents the comment injection mode where both service tags and tracing tags. Tracing tags include span id, trace id and sampling priority.
+	SQLInjectionModeFull SQLCommentInjectionMode = 2
+)
+
+// Key names for SQL comment tags.
+const (
+	sqlCommentKeySamplingPriority = "ddsp"
+	sqlCommentTraceID             = "ddtid"
+	sqlCommentSpanID              = "ddsid"
+	sqlCommentService             = "ddsn"
+	sqlCommentVersion             = "ddsv"
+	sqlCommentEnv                 = "dde"
+)
+
+// SQLCommentCarrier is a carrier implementation that injects a span context in a SQL query in the form
+// of a sqlcommenter formatted comment prepended to the original query text.
+// See https://google.github.io/sqlcommenter/spec/ for more details.
+type SQLCommentCarrier struct {
+	Query  string
+	Mode   SQLCommentInjectionMode
+	SpanID uint64
+}
+
+// Inject injects a span context in the carrier's query.
+func (c *SQLCommentCarrier) Inject(spanCtx ddtrace.SpanContext) error {
+	c.SpanID = random.Uint64()
+	tags := make(map[string]string)
+	switch c.Mode {
+	case SQLInjectionDisabled:
+		return nil
+	case SQLInjectionModeFull:
+		samplingPriority := 0
+		var traceID uint64
+		ctx, ok := spanCtx.(*spanContext)
+		if ok {
+			if sp, ok := ctx.samplingPriority(); ok {
+				samplingPriority = sp
+			}
+			if ctx.TraceID() > 0 {
+				traceID = ctx.TraceID()
+			}
+		}
+		if traceID == 0 {
+			traceID = c.SpanID
+		}
+		tags[sqlCommentTraceID] = strconv.FormatUint(traceID, 10)
+		tags[sqlCommentSpanID] = strconv.FormatUint(c.SpanID, 10)
+		tags[sqlCommentKeySamplingPriority] = strconv.Itoa(samplingPriority)
+		fallthrough
+	case SQLInjectionModeService:
+		ctx, ok := spanCtx.(*spanContext)
+		var env, version string
+		if ok {
+			if e, ok := ctx.meta(ext.Environment); ok {
+				env = e
+			}
+			if v, ok := ctx.meta(ext.Version); ok {
+				version = v
+			}
+		}
+		if globalconfig.ServiceName() != "" {
+			tags[sqlCommentService] = globalconfig.ServiceName()
+		}
+		if env != "" {
+			tags[sqlCommentEnv] = env
+		}
+		if version != "" {
+			tags[sqlCommentVersion] = version
+		}
+	}
+	c.Query = commentQuery(c.Query, tags)
+	return nil
+}
+
+// commentQuery returns the given query with the tags from the SQLCommentCarrier applied to it as a
+// prepended SQL comment. The format of the comment follows the sqlcommenter spec.
+// See https://google.github.io/sqlcommenter/spec/ for more details.
+func commentQuery(query string, tags map[string]string) string {
+	if len(tags) == 0 {
+		return ""
+	}
+	serializedTags := make([]string, 0, len(tags))
+	for k, v := range tags {
+		eKey := url.QueryEscape(k)
+		eKey = strings.Replace(eKey, "+", "%20", -1)
+		sKey := strings.ReplaceAll(eKey, "'", "\\'")
+		eVal := url.QueryEscape(v)
+		eVal = strings.Replace(eVal, "+", "%20", -1)
+		escVal := strings.ReplaceAll(eVal, "'", "\\'")
+		sValue := fmt.Sprintf("'%s'", escVal)
+		serializedTags = append(serializedTags, fmt.Sprintf("%s=%s", sKey, sValue))
```

Comment:
I think this is much better, but would be nice to have some inline guidance on what's going on here. There are way too many cryptic variables.

Here's a simplification.
```suggestion
		// we need to URL-encode all + characters and escape single quotes
		// http://<link-to-description-of-technique>/
		key := url.QueryEscape(strings.Replace(k, "+", "%20", -1))
		key = strings.ReplaceAll(key, "'", "\\'")
		val := url.QueryEscape(strings.Replace(v, "+", "%20", -1))
		val = strings.ReplaceAll(eVal, "'", "\\'")
		serializedTags = append(serializedTags, fmt.Sprintf("%s='%s'", key, val))
```
I think you provided some link with a spec explaining this manoeuvre IIRC? Otherwise ignore that part.

Reply:
This works with a slight modification. It looks like `url.QueryEscape` reverts the swap of `+` by `%20` so I had to swap the order of operation for that one. Thankfully, tests covered this case well.